sbor/src/schema/schema_validation/type_metadata_validation.rs

```diff
@@ -0,0 +1,185 @@
+            // None can apply to any field count
+        }
+        Children::Fields(fields_matadata) => {
+            if fields_matadata.len() != field_count {
+                return Err(SchemaValidationError::TypeMetadataContainedWrongNumberOfChildNames);
+            }
+            for field_metadata in fields_matadata.iter() {
+                let FieldMetadata { field_name } = field_metadata;
+                validate_field_name(field_name)?;
+            }
+        }
+        Children::Variants(_) => {
+            return Err(SchemaValidationError::TypeMetadataForFieldsContainedEnumVariantChildNames)
+        }
+    }
+    Ok(())
+}
+
+pub fn validate_enum_metadata(
+    type_metadata: &TypeMetadata,
+    variants: &BTreeMap<u8, Vec<LocalTypeIndex>>,
+) -> Result<(), SchemaValidationError> {
+    let TypeMetadata {
+        type_name,
+        children,
+    } = type_metadata;
+    validate_type_name(type_name.as_ref())?;
+
+    match &children {
+        Children::None | Children::Fields(_) => {
+            return Err(SchemaValidationError::TypeMetadataForEnumIsNotEnumVariantChildNames)
+        }
+        Children::Variants(variants_metadata) => {
+            if variants_metadata.len() != variants.len() {
+                return Err(SchemaValidationError::TypeMetadataContainedWrongNumberOfChildNames);
+            }
+            for (discriminator, variant_metadata) in variants_metadata.iter() {
+                let Some(child_types) = variants.get(discriminator) else {
+                    return Err(SchemaValidationError::TypeMetadataHasMismatchingEnumDiscriminator)
+                };
+
+                let TypeMetadata {
+                    type_name,
+                    children,
+                } = variant_metadata;
+                validate_enum_variant_name(type_name.as_ref())?;
+                validate_fields_child_names(children, child_types.len())?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn validate_type_name(name: &str) -> Result<(), SchemaValidationError> {
+    if name.len() == 0 {
+        return Err(SchemaValidationError::InvalidTypeName {
+            message: "Type name cannot be empty".into(),
+        });
+    }
+    if name.len() > 100 {
+        return Err(SchemaValidationError::InvalidTypeName {
+            message: "Type name cannot be more than 100 characters".into(),
+        });
+    }
+    for char in name.chars() {
+        if !matches!(char, '0'..='9' | 'A'..='Z' | 'a'..='z' | '_') {
+            return Err(SchemaValidationError::InvalidTypeName {
+                // We need to validate this because we need to generate types from these names
+                // Rust is much less prescriptive for identifier names (see https://doc.rust-lang.org/reference/identifiers.html and https://unicode.org/reports/tr31/)
+                // But we can always loosen this later
+                message: "At present, types names must match [0-9A-Za-z_]+".into(),
+            });
```

Comment:
How about not allowing type name to start with [0-9]? Want this for ease of type generation.

Reply:
Ahh right, yes indeed 👍. Easing type generation was the whole point of this restriction, and yet I still failed :P.

I might also restrict to avoid `_` as the first character?